nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-site-to-site/src/main/java/org/apache/nifi/remote/protocol/AbstractFlowFileServerProtocol.java

```diff
@@ -257,7 +259,13 @@ public int transferFlowFiles(final Peer peer, final ProcessContext context, fina
             session.read(flowFile, new InputStreamCallback() {
                 @Override
                 public void process(final InputStream in) throws IOException {
-                    final DataPacket dataPacket = new StandardDataPacket(toSend.getAttributes(), in, toSend.getSize());
+                    LinkedHashMap<String, String> attributes = new LinkedHashMap<>();
+                    String[] keySet = toSend.getAttributes().keySet().toArray(new String[0]);
+                    Arrays.sort(keySet);
+                    for(String key: keySet){
+                        attributes.put(key, toSend.getAttributes().get(key));
+                    }
```

Comment:
This approach effectively makes an additional copy of the attributes prior to transmission, which introduces some inefficiency when sending packets. It would be better to evaluate the initial attribute storage in the FlowFile implementation, as opposed to making this change.

Reply:
Thank you for your suggestion! Your concern is reasonable which I haven't thought about. I would make more effort to try fixing it by replacing the initial storage of attributes from HashMap to LinkedHashMap. And you are right that since it's a low-level change that may widely influence the entire module, we still need to do more evaluation on whether to change it or just let it remain the original way.